macros/src/lib.rs

```diff
@@ -0,0 +1,42 @@
+extern crate proc_macro;
+
+use proc_macro::TokenStream;
+
+fn get_value_from_token_stream(input: TokenStream) -> String {
+    let val = format!("{}", input);
+    if !val.starts_with('"') || !val.ends_with('"') {
+        panic!("Argument must be a string literal.");
+    }
+    let len = val.len();
+    (&val[1..len - 1]).to_string()
+}
+
+#[proc_macro]
+pub fn tinystr4(input: TokenStream) -> TokenStream {
+    let val = get_value_from_token_stream(input);
+    let bytes: u32 = tinystr::TinyStr4::from_bytes(val.as_bytes())
+        .expect("Failed to construct TinyStr from input")
+        .into();
+    let formula = format!("unsafe {{ tinystr::TinyStr4::new_unchecked({}) }}", bytes);
```

Comment:
What do you think about making this sub-crate only emit the bytes themselves, and wrapping the bytes in a normal `macro_rules` in the main crate, such that you can use `$crate::TinyStr4` for better resolution behavior?

Reply:
I would prefer not, or at least make it a separate macro. The bytes value is an internal detail and I'd prefer to assume that we are always using `Into` and `new_unchecked` paired up without assuming the value in between.

I'm not sure if I understand the value either, can you explain it further?

Reply:
Oh, I'm not suggesting to replace the public macro with the bytes version.  I'm only suggesting to make a bytes macro named something cryptic like `tinystr4_internal_bytes!`, and then define `tinystr4!` in the main tinystr crate using a `macro_rules!` that internally calls `tinystr4_internal_bytes!`.

The proc macro resolves to a literal string of text; the compiler can't do anything smart with it.  For example, does the macro work if you import tinystr under some other name?  `$crate` solves this problem, and I think it can only be used from a regular `macro_rules!`.

Reply:
Why would you want to do that? Why not just use `tinystr4` proc macro all the way?

Reply:
Did you read the second paragraph?

> The proc macro resolves to a literal string of text; the compiler can't do anything smart with it. For example, does the macro work if you import tinystr under some other name? $crate solves this problem, and I think it can only be used from a regular macro_rules!.

Reply:
I did read it, but I'm not sure if I understand it.

`$crate` works in proc macro - https://github.com/zbraniecki/unic-locale/blob/master/unic-langid-macros-impl/src/lib.rs#L18

Is your main concern the naming collision for `tinystr` that is resolvable via `$crate` or is there more?

Reply:
In my very limited experience writing public macros, crate name resolution has been a problem, and it can be solved using $crate.

If we are exporting the macro from the main tinystr crate, then I assume the macro needs to be defined in that crate in order for $crate to work.

Reply:
hmm, can you explain a scenario or sample code that may not work with this code?

Reply:
One might have a private mod called `tinystr` (maybe it contains helper functions or something), and they import the tinystr crate under a different name.  The way you have the macro written, it sticks a literal "tinystr" into the source file, which will look inside the private tinystr mod, instead of the real one that was imported under a different name.

Reply:
This is the problem `$crate` is designed to solve, and it can be used in proc macros. There's no need for gymnastics.